game-core/src/test/java/org/triplea/common/util/concurrent/CompletableFutureUtilsTest.java

```diff
@@ -0,0 +1,47 @@
+package org.triplea.common.util.concurrent;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.triplea.common.util.concurrent.CompletableFutureUtils.logExceptionWhenComplete;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+final class CompletableFutureUtilsTest {
+  @ExtendWith(MockitoExtension.class)
+  @Nested
+  final class LogExceptionWhenCompleteTest {
+    private static final String ERROR_MESSAGE = "error message";
+
+    @Mock
+    private Logger logger;
+
+    @Test
+    void shouldNotWriteLogWhenFutureCompletesNormally() {
+      final CompletableFuture<Object> future = new CompletableFuture<>();
+      logExceptionWhenComplete(future, ERROR_MESSAGE, logger);
+      future.complete(new Object());
+
+      verify(logger, never()).log(any(Level.class), anyString(), any(Throwable.class));
+    }
+
+    @Test
+    void shouldWriteLogWhenFutureCompletesExceptionally() {
+      final CompletableFuture<Object> future = new CompletableFuture<>();
+      logExceptionWhenComplete(future, ERROR_MESSAGE, logger);
+      final Exception ex = new Exception();
+      future.completeExceptionally(ex);
```

Comment:
Yup

Reply:
Hmm, is that really necessary?  `verify(obj)` is equivalent to `verify(obj, times(1))`.  So, if there was more than one `logger.log(Level.SEVERE, ERROR_MESSAGE, ex)` call, the `verify()` assertion would fail.

If the concern is that "something else" may be making the `log()` call we're checking for, I think we're ok, as that would show up as two or more calls and the existing `verify()` assertion will fail.  However, if the concern is that there are "no other" `log()` calls being made by `logExceptionWhenComplete()`, then, yeah, I guess we'd need to add the additional assertion, but I'm not sure if that's overconstraining the implementation.

Is there a different scenario that you're concerned about that I didn't cover?

The only other case I can think of is if `logExceptionWhenComplete()` made the `log()` call directly and not within the `whenComplete()` closure.  I can see how that might make the test pass even though the implementation is incorrect.  However, I'd argue that that scenario may not even be possible because there's no way `logExceptionWhenComplete()` can get access to the `Throwable` instance we're testing for.  Thus, it could never make a `log()` call with the `Throwable` that would match the call arguments we're testing for.  That is, I don't think an implementation could replace the comment below with something that could make the test assertion pass:

```java
static void logExceptionWhenComplete(final CompletableFuture<?> future, final String message, final Logger logger) {
  logger.log(Level.SEVERE, message, /* what to pass here? */);
  future.whenComplete((result, ex) -> {
    // do nothing
  });
}
```

Even if one could somehow pass the correct exception, the other test case (`shouldNotWriteLogWhenFutureCompletesNormally`) would fail, and thus detect the invalid implementation.

Reply:
Leaving it up o you to merge this!

Reply:
:smile: